(feat) O3-4535 Add widget to display diagnosis tags in styleguide

chibongho opened the pull request:
# Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. Ensure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing#contributing-guidelines) label (such as `feat`, `fix`, or `chore`, among others). See existing PR titles for inspiration.

## For changes to apps

- [ ] My work conforms to the [**O3 Styleguide**](https://om.rs/styleguide) and [**design documentation**](https://om.rs/o3ui).

## If applicable

- [ ] My work includes tests or is validated by existing tests.
- [ ] I have updated the [esm-framework mock](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-framework/mock.tsx) to reflect any API changes I have made.

## Summary
<!-- Please describe what problems your PR addresses. -->

This PR adds a configurable component to display diagnoses as Carbon tags, with the tag color configurable based on the diagnoses' ranks (1 = Primary, 2 = Secondary). While the component will only be used in patient chart visit's table / visits summary cards, I think it's a component that is universal enough for other ESMs to use.

Not in this PR, but I intend to also add a similar component for displaying coded obs value as Carbon tags. The configuration of that component will be a bit more complex, to allow for some form of "configuration registration" where we can define multiple configs for the same component. [here](https://talk.openmrs.org/t/proposal-to-add-more-configurable-components-in-o3-esm-core/45168/3). 

However, I didn't make the `<DiagnosisTag>` component as complex; instead, it's just one config that defines how all diagnoses should be displayed. (I think it makes more sense to have one universal config for diagnoses tag colors. For obs, OTOH, it makes sense to have multiple configuration for different concepts.)

## Screenshots
<!-- Required if you are making UI changes. -->
![image](https://github.com/user-attachments/assets/31316743-a916-4bc4-82cd-62d0a52817ec)

## Related Issue
<!-- Paste the link to the Jira ticket here if one exists. -->
<!-- https://issues.openmrs.org/browse/O3- -->

## Other
<!-- Anything not covered above -->
I had trouble testing this component locally in patient-chart, where I intend to use it. I followed the instructions in the README to to `yarn link`, run `yarn run:shell` and override the module URL, but it was not able to get the component. I "tested" it by including this component in the left nav. I'm not sure if I exported it incorrectly.

ibacher commented on packages/framework/esm-styleguide/src/config-schema.ts:
```diff
@@ -1,5 +1,47 @@
+  'outline',
+] as const;
+
+export const diagnosisTagConfigSchema: ConfigSchema = {
```
Could we make the config schema here just:

```json
{
  "primary": "red",
  "secondary": "blue"
}
```

I realize that's not as flexible as the underlying data model, but we've never really used anything other than primary and secondary diagnoses and I'm not sure the "ranks" part is actually well understood.

ibacher commented on packages/framework/esm-styleguide/src/config-schema.ts:
```diff
@@ -1,5 +1,47 @@
-import { Type, validators } from '@openmrs/esm-config';
+import { type ConfigSchema, Type, validators } from '@openmrs/esm-config';
+import { type DiagnosisTagsConfig } from './diagnoses-tags/diagnosis-tags.resource';
```
I think it's probably preferable to have this type defined here.

ibacher commented on packages/framework/esm-styleguide/src/diagnoses-tags/diagnosis-tags.resource.ts:
```diff
@@ -0,0 +1,58 @@
+  appliedToRanks: Array<number>;
+}
+
+export const diagnosisTagConfigSchema: ConfigSchema = {
```
I assume this will go before we merge it in?

ibacher commented on packages/framework/esm-styleguide/src/diagnoses-tags/diagnosis-tags.resource.ts:
```diff
@@ -0,0 +1,58 @@
+    externalModuleName: '@openmrs/esm-styleguide',
+  });
+  const key = '@openmrs/esm-styleguide/diagnosisTags';
+  const { data: rankToColorMap } = useSWRImmutable(config ? key : null, () => {
```
I strongly dislike this use of SWR as a generic cache. I feel like we can more easily just cache the results in an object or map stored outside of the hook and get a better benefit here.

ibacher commented on packages/framework/esm-styleguide/src/utils.ts:
```diff
@@ -0,0 +1,21 @@
+ * Array of valid values for Carbon tag's "type" attribute.
+ * Options from https://react.carbondesignsystem.com/?path=/docs/components-tag--overview
+ */
+export const carbonTagColors = [
```
This also seems duplicative of what we have in the config schema, but in this case, it makes some sense to keep this version and remove it from the config schema.

ibacher reviewed (changes requested):
Thanks @chibongho! I think this is something that makes for a very reasonable component, but there are a few things we should clean-up in this PR before it's merged and I'm hoping we can do something similar for the configuration, since I don't think what's here is going to be very easy for anyone to use.

ibacher commented:
> Have you done it recently, @ibacher?

`yarn link` doesn't appropriately handle `workspace:*`-style version declarations, but removing those declarations makes the linking process _within the repo_ more challenging and that's what we've optimized for. What I've done where I need to do this is manually shift those versions, i.e., replace `workspace:*` with whatever the version in package.json is.

ibacher commented on packages/framework/esm-styleguide/src/index.ts:
```diff
@@ -1,27 +1,28 @@
+export * from './spinner';
+export * from './toasts';
+export * from './workspaces';
+export * from './diagnoses-tags';
```
Shouldn't this be after `./datepicker`?

ibacher commented on packages/framework/esm-styleguide/src/diagnoses-tags/diagnosis-tags.component.tsx:
```diff
@@ -0,0 +1,30 @@
+import { useDiagnosisRankToColorMap } from './diagnosis-tags.resource';
+
+interface DiagnosisTagsProps {
+  diagnoses: Diagnosis[];
```
Style-wise we've preferred this syntax. They're equivalent, but consistency is important.

```suggestion
  diagnoses: Array<Diagnosis>;
```

ibacher commented on packages/framework/esm-styleguide/src/diagnoses-tags/diagnosis-tags.component.tsx:
```diff
@@ -0,0 +1,30 @@
+ * Carbon tags, with colors configured base on the rank (e.g. primary or
+ * secondary) of the diagnoses.
+ */
+export const DiagnosisTags: React.FC<DiagnosisTagsProps> = ({ diagnoses }) => {
```
I think `DiagnosisTags` is the appropriate name for the component. Should we rename the directories to reflect taht?

chibongho commented:
> Thanks @chibongho! I think this is something that makes for a very reasonable component, but there are a few things we should clean-up in this PR before it's merged and I'm hoping we can do something similar for the configuration, since I don't think what's here is going to be very easy for anyone to use.

Addressed the feedback. This definitely simplifies the config a bit.

My next component is for displaying coded obs values as tags. Config will be more complex, but I'll try to model it similarly.
